### DISCUSSION TITLE
Module localization Part 2

### DISCUSSION THREAD

**Discussion by JanOlsmar** [oqtane/oqtane.framework#2303] [Q&A] [answered]
I have still some problem regarding text translation.  I can't get Oqtane components to work for translation in my modules.
For example:
"ActionLink Action="Add" Security="SecurityAccessLevel.Edit"  ResourceKey="Add Point" 
and
     Label Class="col-sm-3" For="latitude" HelpText="Enter latitude" ResourceKey="Latitude">Latitude: </Label>

As it works for the framework I probably do something wrong :)

**Comment by leigh-pointer** [comment]
@JanOlsmar im sure that the Oqtane components need to be translated in the framework translation and not the module.

**Reply by JanOlsmar** [reply]
So using those components in a module is not an option then.
Why use them in Module creator.

**Comment by sbwalker** [comment]
@JanOlsmar you should absolutely be able to use the shared framework components within your localized modules. Let's focus on the most common shared component - the Label component. You should be able to include references to the Label component within your module and include the standard ResourceKey attribute. Then in the corresponding RESX file for the component you would declare the ResourceKey.Text and ResourceKey.HelpText localizations. Then you need to compile the module to produce a satellite assembly which is deployed to the framework /bin folder. When the framework starts up it should load the satellite assembly into the app domain which allows the Label component to load the localized values from the satellite assembly for the module. If this is not happening then we need to investigate the root cause. Are you using the latest code in the dev branch or the last official release version? I am asking because I made some changes recently in the dev branch based on your earlier request to allow an admin to add languages to a site even if a translation does not exist.

**Reply by JanOlsmar** [reply]
I am using 3.1.3 release version. I use Module creator. Change the cmd file as leigh-poiter told me ysterday. A dll is in the sv-SE folder in the client framwork folder. 
Other translations works but not for example the label.

**Reply by JanOlsmar** [reply]
I have tested some more. Regarding "Label" it works  if you use it as in the Edit page (Module Creator) but not in Setting Page for settings for your module.
"ActionLink" can't I get to work for translations.

**Comment by sbwalker** [comment]
@JanOlsmar the Settings component has some additional complexity as compared to "standard" module components, as a "setting" component is essentially a sub-component that is dynamically injected into the standard ModuleSettings.razor module component. As a result of this hierarchical relationship there may be an issue related to how it loads its localized resources. I will will have to investigate further... but thank you for providing the additional info that this only seems to affect Settings components.

**Reply by JanOlsmar** [reply]
The ActionLink issue is still valid and maybe more controls. I will know better when I test more.

**Comment by sbwalker** [comment, marked as answer]
@JanOlsmar are you sure you are using the ActionLink correctly? You MUST specify the ResourceKey attribute 

```
<ActionLink Action="Add" Text="Add Role" ResourceKey="AddRole" />
```

and you must specify the Name in the RESX file correctly (ie. in the format {ResourceKey}.{ComponentPropertyName}):

Name= "AddRole.Text"
Value = "Add Role"

**Reply by JanOlsmar** [reply]
Thanks :) No I had not. I had followed the pattern used in Module Creator. So now ActionLink works fine too.  Maybe Module Creator should change so no more get fooled. (I love the creator to start a new module)

**Reply by sbwalker** [reply]
@JanOlsmar thank you for letting me know about the missing attribute in the External Module template - this can be resolved very easily to prevent confusion in the future.

**Reply by sbwalker** [reply]
@JanOlsmar #2331 resolves this issue where the ActionLink and ActionDialog keys were not included in the proper format in the RESX file

**Comment by sbwalker** [comment]
@JanOlsmar in regards to the Settings issue you reported earlier in this thread, I confirmed that is does have some additional complexity as compared to "standard" module components, as a "setting" component is essentially a sub-component that is dynamically injected into the core Settings.razor module component. If you look at the Settings.razor for the default Html/Text Module you will see that it uses a slightly modified approach to resolve this issue:

https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Client/Modules/HtmlText/Settings.razor 

Notice the use of the ResourceType attribute in the Label component declaration. This property can be used to explicitly specify the type name of the component being localized, and it is necessary in this scenario because the parent module type is Oqtane.Admin.Modules.Settings and the Html/Text module settings component is being dynamically injected into it. If you do not like the hardcoding of the type name you can probably get it programmatically using the following code:

```
	protected override void OnInitialized()
	{
		 resourceType = GetType().FullName;
	}
```

**Reply by JanOlsmar** [reply]
Tryed both methods. I could't get it work.

**Reply by JanOlsmar** [reply]
This works. Make it simple principe :)
 <Label Class="col-sm-3" For="value" HelpText="@Localizer["Helptext"]">@Localizer["Key"]</Label>

**Reply by hishamco** [reply]
No need to use localizer here, you could use `ResourceKey` instead

**Reply by JanOlsmar** [reply]
@hishamco AS I wrote I tryed  the methods Shaun sugested.  I could't get it work.

**Reply by hishamco** [reply]
I saw you marked the answer, I though this is an extra thing, anyhow try to check the resx file location

**Reply by JanOlsmar** [reply]
@hishamco Try to get it work. If you can get it to work please let me know.  Until then I will use my method in the Settings.

**Reply by hishamco** [reply]
Do you have a minimal repo to reproduce?

**Reply by JanOlsmar** [reply]
@hishamco If you like to test, just make a module with ccreator and test translate the label in Settings.

**Comment by sbwalker** [comment]
@JanOlsmar the ResourceType must be specified as a fully qualified type name which includes the class and assembly. Is it possible that you did not include the proper assembly specification (as modules use a different naming convention than the core framework? I have updated the external module template to include ResourceType in the Settings.razor which is generated.

private string resourceType = "[Owner].[Module].Settings, [Owner].[Module].Client.Oqtane"; // for localization

**Reply by JanOlsmar** [reply]
You are so right. I tested a few combinations :) But all used Oqtane.Client instead of Client.Oqtane the Html module fooled me. It's easy to fool me. Very good that it is included in the template. Thanks.
